pkg/modeling/modeling.go

```diff
@@ -0,0 +1,340 @@
+	// Only when the resourceLists are exactly the same can they be counted as the same node.
+	// +required
+	quantity int
+
+	// resourceList records the resource list of this node.
+	// It maybe contain cpu, mrmory, gpu...
+	// User can specify which parameters need to be included before the cluster starts
+	// +required
+	resourceList corev1.ResourceList
+}
+
+func InitSummary(rsName []corev1.ResourceName, rsList []corev1.ResourceList) (modelingSummary, error) {
+	if len(rsName) != 0 && len(rsList) != 0 && (len(rsName) != len(rsList[0])) {
+		return nil, errors.New("the number of resourceName is not equal the number of resourceList")
+	}
+	var ms modelingSummary
+	if len(rsName) != 0 {
+		DefaultModelSorting = rsName
+	}
+	if len(rsList) != 0 {
+		DefaultModel = rsList
+		defaultModelLevel = len(rsList)
+		ms = make(modelingSummary, defaultModelLevel)
+	} else {
+		ms = make(modelingSummary, defaultModelLevel)
+	}
+	// generate a sorted array by first priority of ResourceName
+	for index := 0; index < len(rsList); index++ {
+		tmpQuantity := rsList[index][rsName[0]]
+		quantityNum, ok := tmpQuantity.AsInt64()
+		if !ok {
+			klog.Infof("Unable to parse the values of %v's quantity in the cluster", rsName[0])
+		}
+		modelSorting = append(modelSorting, quantityNum)
+	}
+	return ms, nil
+}
+
+func (ms *modelingSummary) getIndex(crn clusterResourceNode) int {
+	tmpQuantity := crn.resourceList[DefaultModelSorting[0]]
+	quantityNum, ok := tmpQuantity.AsInt64()
+	if !ok {
+		klog.Infof("Unable to parse the values of %v's quantity in the cluster", crn.resourceList)
+	}
+	index := searchLastLessElement(modelSorting, quantityNum)
+
+	// the last element represent the +∞
+	if index == len(modelSorting)-1 {
+		return index
+	}
+	return index + 1
+}
+
+func searchLastLessElement(nums []int64, target int64) int {
+	low, high := 0, len(nums)-1
+	for low <= high {
+		mid := low + ((high - low) >> 1)
+		if nums[mid] <= target {
+			if (mid == len(nums)-1) || (nums[mid+1] > target) {
+				// find the last less element that equal to target element
+				return mid
+			}
+			low = mid + 1
+		} else {
+			high = mid - 1
+		}
+	}
+	return -1
+}
+
+// clusterResourceNodeComparator provides a fast comparison on clusterResourceNodes
+func clusterResourceNodeComparator(a, b interface{}) int {
+	s1 := a.(clusterResourceNode)
+	s2 := b.(clusterResourceNode)
+	var diff int64
+	for index := 0; index < len(DefaultModelSorting); index++ {
+		tmp1, tmp2 := s1.resourceList[DefaultModelSorting[index]], s2.resourceList[DefaultModelSorting[index]]
+		quantity1, ok := tmp1.AsInt64()
+		if !ok {
+			klog.Infof("ModelComparator: Unable to parse the values of %v's quantity1 in the cluster", DefaultModelSorting[index])
+		}
+		quantity2, ok := tmp2.AsInt64()
+		if !ok {
+			klog.Infof("ModelComparator: Unable to parse the values of %v's quantity2 in the cluster", DefaultModelSorting[index])
+		}
+		diff = quantity1 - quantity2
+		if diff < 0 {
+			return -1
+		}
+		if diff > 0 {
```

Comment:
There is something wrong with the comparison function. For example, if a's limit is 2C 5G, b is 1C 8G, that will be CPU diff > 0, memory diff < 0. So I think should not  return directly.

Reply:
I also want to discuss this part with you. I'm comparing the quota by priority. Assuming that the priority of the cpu is high, the cpu is small means that this resource quota's priority is low.

But now there is a situation, as you said. If cpu is small, mem is large. So should this be considered big or small? How do you think about this situation?

Reply:
Could you share your slack with me? we can talk about it on slack.